### PR TITLE
fix(collection): only send bypassDocumentValidation if true

### PR DESCRIPTION
**Pull request by shsaskin**
The bypassDocumentValidation key will only be set if it explicitly
receives a true, otherwise it remains undefined.

Fixes NODE-1492

### CODE REVIEW

File: lib/bulk/unordered.js [hunk 1]
```diff
@@ -161,6 +161,8 @@ FindOperatorsUnordered.prototype.remove = function() {
 // Add to the operations list
 //
 var addToOperationsList = function(_self, docType, document) {
+  console.log('here is document: ');
```
Comment (kvwalker) [inline lib/bulk/unordered.js]: Were these comments for testing?

File: lib/apm.js [hunk 2]
```diff
@@ -9,7 +9,7 @@ class Instrumentation extends EventEmitter {
   instrument(MongoClient, callback) {
     // store a reference to the original functions
     this.$MongoClient = MongoClient;
-    const $prototypeConnect = this.$prototypeConnect = MongoClient.prototype.connect; // eslint-disable-line
+    const $prototypeConnect = (this.$prototypeConnect = MongoClient.prototype.connect); // eslint-disable-line
```
Comment (kvwalker) [inline lib/apm.js]: This was probably just changed when you ran `npm run format`, but I'm okay with leaving this fix if @daprahamian is! If you do leave this, remove the comment.

File: lib/bulk/unordered.js (hunk 1 above)
Comment (shsaskin) [reply lib/bulk/unordered.js]: Yes, I must have missed them.

File: lib/apm.js (hunk 2 above)
Comment (daprahamian) [reply lib/apm.js]: Since it has been annoying us from the start, I say we leave it in, and remove the eslint comment.

File: lib/collection.js [hunk 3]
```diff
@@ -581,7 +581,6 @@ var bulkWrite = function(self, operations, options, callback) {
     options.ordered === true || options.ordered == null
       ? self.initializeOrderedBulkOp(options)
       : self.initializeUnorderedBulkOp(options);
-
```
Comment (daprahamian) [inline lib/collection.js]: unnecessary whitespace change

File: lib/collection.js [hunk 4]
```diff
@@ -2225,12 +2224,13 @@ var findAndModify = function(self, query, sort, doc, options, callback) {
   var finalOptions = applyWriteConcern(options, { db: self.s.db, collection: self }, options);
 
   // Decorate the findAndModify command with the write Concern
+
```
Comment (daprahamian) [inline lib/collection.js]: unnecessary whitespace change

File: lib/collection.js [hunk 5]
```diff
@@ -2867,7 +2867,7 @@ var mapReduce = function(self, map, reduce, options, callback) {
   };
 
   // Exclusion list
-  var exclusionList = ['readPreference', 'session'];
+  var exclusionList = ['readPreference', 'session', 'bypassDocumentValidation'];
```
Comment (daprahamian) [inline lib/collection.js]: why does this need to be added to the exclusion list?

File: test/unit/bypass_validation_tests.js [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+  const test = {};
+  beforeEach(() => {
+    return mock.createServer().then(server => {
+      test.server = server;
+    });
+  });
+  afterEach(() => mock.cleanup());
+
+  // aggregate
+  it('should only set bypass document validation if strictly true in aggregate', function(done) {
```
Comment (daprahamian) [inline test/unit/bypass_validation_tests.js]: A lot of the code in this file is repeated. Could we refactor it to be more DRY?

File: lib/collection.js (hunk 5 above)
Comment (shsaskin) [reply lib/collection.js]: I added it to the exclusion list because the loop after updates all the keys in mapCommandHash to be the options passed in, so when false is passed in, it will update bypassDocumentValidation to be false instead of undefined. I put byypassDocumentValidation in the exclusion list to prevent that. Should I delete bypassDocumentValidation from options or mapCommandHash if it's false instead?

Comment (daprahamian) [reply lib/collection.js]: ah, that makes sense 👍

File: test/unit/bypass_validation_tests.js [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+      if (doc.ismaster) {
+        request.reply(Object.assign({}, mock.DEFAULT_ISMASTER));
+      } else if (doc.endSessions) {
+        request.reply({ ok: 1 });
+      }
+    });
+
+    client.connect(function(err, client) {
+      expect(err).to.not.exist;
+      var db = client.db('test');
```
Comment (daprahamian) [inline test/unit/bypass_validation_tests.js]: replace the use of `var` with `const` where possible, `let` otherwise.

File: test/unit/bypass_validation_tests.js [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+      var collection = db.collection('test_c');
+
+      var options = {
```
Comment (daprahamian) [inline test/unit/bypass_validation_tests.js]: this can be abbreviated to

```js
const options = { bypassDocumentValidation };
```

File: test/unit/bypass_validation_tests.js [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+        {
+          $project: {}
+        }
+      ];
+      collection.aggregate(pipeline, options).next(() => close());
+    });
+  }
+  // aggregate
+  it('should only set bypass document validation if strictly true in aggregate', function(done) {
+    testAggregate(true, done);
```
Comment (daprahamian) [inline test/unit/bypass_validation_tests.js]: Instead of passing in a boolean, you can pass in an object:

```js
testAggregate({ expected: true, actual: true }, done);
```

This is also true for the bulkWrite tests below, where you are passing a flag for ordered.

File: test/unit/bypass_validation_tests.js [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+  it('should only set bypass document validation if strictly true in findAndModify', function(done) {
+    testFindAndModify({ expected: true, actual: true }, done);
+  });
+
+  it('should not set bypass document validation if not strictly true in findAndModify', function(done) {
+    testFindAndModify({ expected: undefined, actual: false }, done);
+  });
+
+  // general test for BlukWrite to test changes made in ordered.js and unordered.js
+  function testBulkWrite(bypassDocumentValidation, done) {
```
Comment (daprahamian) [inline test/unit/bypass_validation_tests.js]: Nice! One last thing: rename the `bypassDocumentValidation` parameter to `config`.
